Support for MERGE (a)-[:T]->(b) in cost planner.

pontusmelke opened the pull request.

pontusmelke committed:
Support for MERGE (a)-[:T]->(b) in cost planner.

pontusmelke commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:
```diff
@@ -929,9 +907,9 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
-    assertNumberOfEagerness(query, 0)
+    assertStats(updateWithBothPlanners(query), relationshipsCreated = 3, labelsAdded = 1)
+    //TODO this we need to consider not only overlap but also what known labels the node we set has
+    assertNumberOfEagerness(query, 1)
```
This is a regression compared to, rule we need to be a little smarter here ( we already have the same regression already merged in the code base around queries like:
`MATCH (a:Foo), (b:Bar)  SET b:Foo` which will be eager in cost but not in rule.

Mats-SX replied:
This query does not make sense. `ON MATCH` only works for `MERGE`.

pontusmelke replied:
I updated my badly copy pasted query in the comment. What is said is still true, the regression is already there on these types of queries)

pontusmelke commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:
```diff
@@ -941,9 +919,9 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
 
-    assertStats(executeWithRulePlanner(query), relationshipsCreated = 3, labelsAdded = 1)
+    assertStats(updateWithBothPlanners(query), relationshipsCreated = 3, labelsAdded = 1)
     assertNumberOfEagerness(query, 1)
```
same here

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/plans/rewriter/unnestOptional.scala:
```diff
@@ -31,13 +31,13 @@ case object unnestOptional extends Rewriter {
     case apply@Apply(lhs,
       Optional(
-      e@Expand(_: Argument, _, _, _, _, _, _))) =>
+      e@Expand(_: Argument, _, _, _, _, _, ExpandAll))) =>
```
This is something that maybe needs to be discussed, but I don't believe this is safe to do when `ExpandInto`

thobe commented:
What does the curly bracket mean in `(a)-[:T}->(b)`? (it's in the commit message)

pontusmelke commented:
@thobe it means that I can't type

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala:
```diff
@@ -128,7 +128,7 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
 
     // WHEN
-    executeWithRulePlanner(query)
+    updateWithBothPlanners(query)
```
We should move this test case to `Merge*AcceptanceTest` (node or relationship version).

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala:
```diff
@@ -153,7 +153,7 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
 
     // WHEN
-    executeWithRulePlanner(query)
+    updateWithBothPlanners(query)
```
We should move this test case to `Merge*AcceptanceTest` (node or relationship version).

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala:
```diff
@@ -183,22 +183,22 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
 
     // WHEN
-    val result = executeWithRulePlanner(query)
+    val result = updateWithBothPlanners(query)
```
We should move this test case to `Merge*AcceptanceTest` (node or relationship version).

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala:
```diff
@@ -698,4 +675,11 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
   }
+
+  test("unwind combined with merge") {
+    val query = "UNWIND [1,2,3,4] AS int MERGE (n {id: int})"
```
We should add a `count(*)` to the end of the query and assert on it.

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeRelationshipAcceptanceTest.scala:
```diff
@@ -19,69 +19,55 @@
 
     // when
-    val r = executeScalar("MATCH (a {name:'A'}), (b {name:'B'}) MERGE (a)-[r:TYPE]->(b) RETURN r").asInstanceOf[Relationship]
+    val result = updateWithBothPlanners("MATCH (a {name:'A'}), (b {name:'B'}) MERGE (a)-[r:TYPE]->(b) RETURN count(*)")
```
Missing non-stats assertion on `result`.

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeRelationshipAcceptanceTest.scala:
```diff
@@ -89,214 +75,169 @@ class MergeRelationshipAcceptanceTest extends ExecutionEngineFunSuite with Query
 
     // when
-    val r = executeScalar("MATCH (a {name:'A'}), (b {name:'B'}) MERGE (a)-[r:TYPE {name:'r2'}]->(b) RETURN r").asInstanceOf[Relationship]
+    val result = updateWithBothPlanners("MATCH (a {name:'A'}), (b {name:'B'}) MERGE (a)-[r:TYPE {name:'r2'}]->(b) RETURN count(r)")
```
Missing assertion on result

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeRelationshipAcceptanceTest.scala:
```diff
@@ -89,214 +75,169 @@ class MergeRelationshipAcceptanceTest extends ExecutionEngineFunSuite with Query
-      r.getType.name() should equal("TYPE")
-    }
+    assertStats(result, relationshipsCreated = 1, propertiesSet = 1)
+    executeWithAllPlanners("MATCH (a {name:'A'})-[r:TYPE {name:'r2'}]->(b {name:'B'}) RETURN a.name").toList should have size 1
```
Assertion could be more detailed (re: above I missed this inlining -- but it could be more explicit than just size check).

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeRelationshipAcceptanceTest.scala:
```diff
@@ -89,214 +75,169 @@ class MergeRelationshipAcceptanceTest extends ExecutionEngineFunSuite with Query
   }
 
-  test("should_handle_on_match") {
+  test("should handle on match") {
```
Missing test case for combining `ON MATCH` and `ON CREATE`

pontusmelke commented:
fixed @Mats-SX s comments.

You get to decide if the `NOT READY TOO MERGE` should go or not. There is regression using `SET` as mentioned above, but that regression is already in the code and could be dealt with as a separate task. The only other thing is the change to `unnestOptional`, if that should be backported or not.

pontusmelke committed:
Fixed bug in OptionalExpandIntoPipe

systay commented:
I've added some comments to the last commit here: https://github.com/pontusmelke/neo4j/commit/5cb5d71d2d5c23f6e8238dfeb7b508ae2c63bf70

pontusmelke committed:
unnestOptional is not safe for merge

A merge relationship must be able to read its own writes,
it is therefore not safe to unnest to OptionalExpand when
in these cases.

pontusmelke committed:
Plan description should print the right thing

pontusmelke committed:
Updated license headers

pontusmelke commented:
@systay I moved the fix to the commit where it belonged and removed the `println`

henriknyman committed:
Fix compilation error
